ragna/core/_components.py

```diff
@@ -76,6 +96,119 @@ def _protocol_model(cls) -> Type[pydantic.BaseModel]:
+# See https://github.com/Quansight/ragna/pull/354#discussion_r1526235318
+@dataclass
+class Embedding:
+    embedding: list[float]
+    chunk: Chunk
+
+    def __init__(self, embedding: list[float], chunk: Chunk):
+        super().__init__()
+        self.embedding = embedding
+        self.chunk = chunk
+
+
+T = TypeVar("T")
+
+
+# The function is adapted from more_itertools.windowed to allow a ragged last window
+# https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.windowed
+def _windowed_ragged(
+    iterable: Iterable[T], *, n: int, step: int
+) -> Iterator[tuple[T, ...]]:
+    window: Deque[T] = deque(maxlen=n)
+    i = n
+    for _ in map(window.append, iterable):
+        i -= 1
+        if not i:
+            i = step
+            yield tuple(window)
+
+    if len(window) < n:
+        yield tuple(window)
+    elif 0 < i < min(step, n):
+        yield tuple(window)[i:]
+
+
+class EmbeddingModel(Component, ABC):
+    _EMBEDDING_DIMENSIONS: int
+    def __init__(self):
+        import tiktoken
+        self._tokenizer = tiktoken.get_encoding("cl100k_base")
+
+    def _chunk_pages(
+            self, pages: Iterable[Page], document_id: UUID, *, chunk_size: int, chunk_overlap: int
+    ) -> Iterator[Chunk]:
+        for window in _windowed_ragged(
+                (
+                        (tokens, page.number)
+                        for page in pages
+                        for tokens in self._tokenizer.encode(page.text)
+                ),
+                n=chunk_size,
+                step=chunk_size - chunk_overlap,
+        ):
+            tokens, page_numbers = zip(*window)
+            yield Chunk(
+                text=self._tokenizer.decode(tokens),  # type: ignore[arg-type]
+                document_id=document_id,
+                page_numbers=list(filter(lambda n: n is not None, page_numbers))
+                             or None,
+                num_tokens=len(tokens),
+            )
+
+    @classmethod
+    def _page_numbers_to_str(cls, page_numbers: Optional[Iterable[int]]) -> str:
+        if not page_numbers:
+            return ""
+
+        page_numbers = sorted(set(page_numbers))
+        if len(page_numbers) == 1:
+            return str(page_numbers[0])
+
+        ranges_str = []
+        range_int = []
+        for current_page_number, next_page_number in itertools_pairwise(
+                itertools.chain(sorted(page_numbers), [None])
+        ):
+            current_page_number = cast(int, current_page_number)
+
+            range_int.append(current_page_number)
+            if next_page_number is None or next_page_number > current_page_number + 1:
+                ranges_str.append(
+                    ", ".join(map(str, range_int))
+                    if len(range_int) < 3
+                    else f"{range_int[0]}-{range_int[-1]}"
+                )
+                range_int = []
+
+        return ", ".join(ranges_str)
+
+    @classmethod
+    def _take_sources_up_to_max_tokens(
```

Comment:
Shouldn't this stay on the VDB base class? An `EmbeddingModel` has no touching points with the `Source`s.